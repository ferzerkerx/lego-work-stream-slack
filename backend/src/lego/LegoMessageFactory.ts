import { SlackMessage } from 'botkit';
import { DateUtils } from '../utils/DateUtils';

export class LegoMessageFactory {
  static defaultConfiguration(channelName: string): TeamChannelConfiguration {
    return {
      channelName: channelName,
      actionDescriptors: [
        { name: 'green', text: 'Green' },
        { name: 'red', text: 'Red' },
        { name: 'orange', text: 'Orange' },
        { name: 'black', text: 'Black' },
      ],
      min: 0,
      max: 8,
    };
  }

  static createMessage(
    config: TeamChannelConfiguration,
    date: Date = new Date()
  ): SlackMessage {
    return {
      text: `Please select your legos for today: ${DateUtils.toPrettyDate(
        date.toDateString()
      )}`,
      attachments: [
        {
          text: 'Choose your legos!',
          fallback: 'Unable to set lego stats',
          callback_id: 'lego_stats',
          color: '#3AA3E3',
          attachment_type: 'default',
          actions: this._createActions(config),
        },
      ],
      channel: config.channelName,
    };
  }

  private static _createActions(config: TeamChannelConfiguration): Array<any> {
    let options: Array<any> = [];
    for (let i: number = config.min; i <= config.max; i++) {
      options.push({
        text: `${i}`,
        value: i,
      });
    }
    return config.actionDescriptors.map(descriptor =>
      this._legoAction(descriptor, options)
    );
  }

  private static _legoAction(
    descriptor: LegoMessageActionDescriptor,
    options: Array<any>
  ): any {
    return {
      name: `lego-select-option-${descriptor.name}`,
      text: descriptor.text,
      type: 'select',
      options: options,
    };
  }
}

export class TeamChannelConfiguration {
  actionDescriptors: LegoMessageActionDescriptor[] = [];
  channelName: string;
  date?: Date = new Date();
  min?: number = 0;
  max?: number = 8;
  cronExpression?: string;
}

export class LegoMessageActionDescriptor {
  name: string;
  text: string;
  description?: string;
  icon?: string;
}
