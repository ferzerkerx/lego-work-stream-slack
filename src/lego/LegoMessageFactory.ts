import { SlackMessage, Storage } from 'botkit';
import { DateUtils } from '../utils/DateUtils';
import { ErrorUtil } from '../utils/ErrorUtil';

export class LegoMessageFactory {
  static defaultConfiguration(): LegoMessageConfig {
    return {
      actionDescriptors: [
        { name: 'green', text: 'Green' },
        { name: 'red', text: 'Red' },
        { name: 'orange', text: 'Orange' },
        { name: 'black', text: 'Black' },
      ],
      min: 0,
      max: 8,
    };
  }

  static createMessage(
    storage: Storage<LegoMessageConfig>,
    channelName: string,
    date: Date = new Date()
  ): Promise<SlackMessage> {
    const conditions: any = [{ channelName: channelName }];

    const query = {
      $and: conditions,
    };

    // @ts-ignore
    return storage.team_configurations
      .find(query)
      .then(storedConfig => {
        let config: LegoMessageConfig =
          storedConfig && storedConfig.length > 0
            ? storedConfig[0]
            : this.defaultConfiguration();

        return {
          text: `Please select your legos for today: ${DateUtils.toPrettyDate(
            date.toDateString()
          )}`,
          attachments: [
            {
              text: 'Choose your legos!',
              fallback: 'Unable to set lego stats',
              callback_id: 'lego_stats',
              color: '#3AA3E3',
              attachment_type: 'default',
              actions: this._createActions(config),
              channelName: channelName,
            },
          ],
        };
      })
      .catch(e => ErrorUtil.defaultErrorHandling(e));
  }

  private static _createActions(config: LegoMessageConfig): Array<any> {
    let options: Array<any> = [];
    for (let i: number = config.min; i <= config.max; i++) {
      options.push({
        text: `${i}`,
        value: i,
      });
    }
    return config.actionDescriptors.map(descriptor =>
      this._legoAction(descriptor, options)
    );
  }

  private static _legoAction(
    descriptor: LegoMessageActionDescriptor,
    options: Array<any>
  ): any {
    return {
      name: `lego-select-option-${descriptor.name}`,
      text: descriptor.text,
      type: 'select',
      options: options,
    };
  }
}

export class LegoMessageConfig {
  actionDescriptors: LegoMessageActionDescriptor[] = [];
  channelName?: string;
  date?: Date = new Date();
  min?: number = 0;
  max?: number = 8;
}

export class LegoMessageActionDescriptor {
  name: string;
  text: string;
  description?: string;
  icon?: string;
}
