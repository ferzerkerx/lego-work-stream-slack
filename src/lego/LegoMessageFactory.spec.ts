import { LegoMessageConfig, LegoMessageFactory } from './LegoMessageFactory';

describe('LegoMessageFactory', () => {
  test('should create message with valid structure', () => {
    // @ts-ignore
    const storage: Storage<LegoMessageConfig> = {
      team_configurations: {
        find: query => {
          return Promise.resolve({
            actionDescriptors: [
              { name: 'green', text: 'Green' },
              { name: 'red', text: 'Red' },
              { name: 'orange', text: 'Orange' },
              { name: 'black', text: 'Black' },
            ],
            min: 0,
            max: 8,
          });
        },
      },
      catch: error => {
        console.log(error);
      },
    };

    LegoMessageFactory.createMessage(
      storage,
      'someChannel',
      new Date('2018-10-01')
    ).then(message => {
      expect(message).toMatchSnapshot();
    });
  });
});
