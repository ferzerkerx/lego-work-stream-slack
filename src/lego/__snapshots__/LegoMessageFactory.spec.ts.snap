// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegoMessageFactory should create message with valid structure 1`] = `
Object {
  "attachments": Array [
    Object {
      "actions": Array [
        Object {
          "name": "lego-select-option-green",
          "options": Array [
            Object {
              "text": "0",
              "value": 0,
            },
            Object {
              "text": "1",
              "value": 1,
            },
            Object {
              "text": "2",
              "value": 2,
            },
            Object {
              "text": "3",
              "value": 3,
            },
            Object {
              "text": "4",
              "value": 4,
            },
            Object {
              "text": "5",
              "value": 5,
            },
            Object {
              "text": "6",
              "value": 6,
            },
            Object {
              "text": "7",
              "value": 7,
            },
            Object {
              "text": "8",
              "value": 8,
            },
          ],
          "text": "Green",
          "type": "select",
        },
        Object {
          "name": "lego-select-option-red",
          "options": Array [
            Object {
              "text": "0",
              "value": 0,
            },
            Object {
              "text": "1",
              "value": 1,
            },
            Object {
              "text": "2",
              "value": 2,
            },
            Object {
              "text": "3",
              "value": 3,
            },
            Object {
              "text": "4",
              "value": 4,
            },
            Object {
              "text": "5",
              "value": 5,
            },
            Object {
              "text": "6",
              "value": 6,
            },
            Object {
              "text": "7",
              "value": 7,
            },
            Object {
              "text": "8",
              "value": 8,
            },
          ],
          "text": "Red",
          "type": "select",
        },
        Object {
          "name": "lego-select-option-orange",
          "options": Array [
            Object {
              "text": "0",
              "value": 0,
            },
            Object {
              "text": "1",
              "value": 1,
            },
            Object {
              "text": "2",
              "value": 2,
            },
            Object {
              "text": "3",
              "value": 3,
            },
            Object {
              "text": "4",
              "value": 4,
            },
            Object {
              "text": "5",
              "value": 5,
            },
            Object {
              "text": "6",
              "value": 6,
            },
            Object {
              "text": "7",
              "value": 7,
            },
            Object {
              "text": "8",
              "value": 8,
            },
          ],
          "text": "Orange",
          "type": "select",
        },
        Object {
          "name": "lego-select-option-black",
          "options": Array [
            Object {
              "text": "0",
              "value": 0,
            },
            Object {
              "text": "1",
              "value": 1,
            },
            Object {
              "text": "2",
              "value": 2,
            },
            Object {
              "text": "3",
              "value": 3,
            },
            Object {
              "text": "4",
              "value": 4,
            },
            Object {
              "text": "5",
              "value": 5,
            },
            Object {
              "text": "6",
              "value": 6,
            },
            Object {
              "text": "7",
              "value": 7,
            },
            Object {
              "text": "8",
              "value": 8,
            },
          ],
          "text": "Black",
          "type": "select",
        },
      ],
      "attachment_type": "default",
      "callback_id": "lego_stats",
      "channelName": "someChannel",
      "color": "#3AA3E3",
      "fallback": "Unable to set lego stats",
      "text": "Choose your legos!",
    },
  ],
  "text": "Please select your legos for today: 2018-10-01",
}
`;
